src/Hazelcast.Net/Serialization/Compact/ReflectionSerializer.cs

```diff
@@ -24,527 +25,634 @@
-            return r;
+            var fieldName = parameters[i].Name;
+            var fieldType = parameters[i].ParameterType;
+            ctorFields.Add(fieldName);
+            values[i] = GetReader(fieldType)(reader, fieldName);
         }
 
-        private static void VerifyTypeIsSupported(object o)
+        // and invoke the constructor with the parameters
+        try
+        {
+            return ctor.Invoke(values);
+        }
+        catch (Exception e)
         {
-            // for FullName to be null, the type would need to be derived from an open generic somehow,
-            // which makes no sense since it is the actual type of a concrete object. we can assume that
-            // type is not going to be null.
-            var type = o.GetType();
-            var name = type.FullName!;
-
-            if (name.StartsWith("<", StringComparison.Ordinal))
-                throw new SerializationException($"The {type} type cannot be serialized via zero-configuration "
-                                                 + "Compact serialization because anonymous types are not supported.");
-
-            if (NonSupportedNamespaces.Any(x => name.StartsWith(x, StringComparison.Ordinal)))
-                throw new SerializationException($"The {name} type is not supported by zero-configuration Compact "
-                                                 + "serialization. Consider writing a custom ICompactSerializer for this type.");
+            throw new SerializationException($"Failed to create an instance of type {type} (Invoking the constructor with '{string.Join(", ", parameters.Select(x => x.ParameterType))}' parameters has thrown, see inner exception).", e);
         }
+    }
+
+    #endregion
+
+    #region Write
+
+    /// <inheritdoc />
+    public virtual void Write(ICompactWriter writer, object obj)
+    {
+        if (obj == null) throw new ArgumentNullException(nameof(obj));
 
-        // for now, we do *not* support all System.* namespaces
-        // we may want to add more later on
-        private static readonly string[] NonSupportedNamespaces = { "System." };
+        VerifyTypeIsSupported(obj);
+
+        foreach (var property in GetProperties(obj.GetType()))
+        {
+            if (!GetValidFieldName(writer, property.PropertyType, property.Name, out var fieldName)) 
+                continue;
+
+            GetWriter(property.PropertyType)(writer, fieldName, property.GetValue(obj));
+        }
+    }
+
+    private static bool GetValidFieldName(ICompactWriter iwriter, Type propertyType, string propertyName, [NotNullWhen(true)] out string? fieldName)
+    {
+        // writer here can be a true CompactWriter, which holds a schema, and is able to provide
+        // 'valid' field names (mostly mapping different casing in names e.g. 'Name' to 'name'),
+        // but it also can be a SchemaBuilderWriter in which case we are writing the schema, and
+        // have to accept the property names as field names.
```

Comment:
I assume field names are counterparts of C# properties on compact and Java side, right? It doesn't relate C# fields.

Reply:
if you write your own serializer then the field name is whatever you specify when you do `WriteInt32("fieldname", value)` - OTOH if you rely on the reflection serializer things start to get different. if you do not provide a schema, we'll create one, and then if the C# property is named `Value` the field will be named `Value`. if you *do* provide a schema we'll try to match a C# property named `Value` with a field named `Value` *or* `value` - it's best effort really.

as soon as you want true reliable interop with Java you should probably write your own serializer anyways.